Provide documentation that is easy to browse online (from github)

f18m opened the pull request:
Hi,
One of the things I find inconvenient about google perftools is the documentation; one of the main reason is that it's not really available online easily (there are old sites caching it but you never know if the github version is newer or not)... that seems a bit weird nowadays :)

Anyway I did a little experiment forking gperftools and then:
1) converting main README and other HTML files under docs/ folder in MarkDown syntax.
    This is useful to have Github able to render the documentation in a nice, eye-candy format

2) linking doc pages to the main README. This makes it easy to browse the docs online.

3) provide a Makefile that allows for offline conversion from MarkDown to HTML. This is useful in case anybody forks the repo and wants to browse documentation locally with his browser.

Can you take a look at  https://github.com/f18m/gperftools  and see if the new documentation format looks nice to you?

To get it into the main repo it should be enough to merge this PR

f18m commented:
One more note: if you think this approach is good and gperftools maintainers are willing to merge it back in main repo, I can complete the work I started (I left out for now cpu profiling docs which are the ones I'm less interested to) :)

f18m commented:
I have further refined the process for HTML generation from Markdown and integrated it with Autoconf. 
I hope this work is found useful and will help increase gperftools visibility (markdown rendered by github should be much easier to search in search engines from developers).

Only cpuprofiler docs is left out... waiting to hear from project maintainers what they think first.

alk commented:
Much thanks for this work. I personally dislike markdown a bit, but since you offer this work and it is improvement over what we have, I'll take it.

But lets make it work first. I've just accepted another documentation fix and your work now needs rebasing :(

Larger issue is making sure it works w.r.t. building source tarballs (i.e. make dist) _and_ building from those tarballs. I don't see anything that tells automake to include your md files into source tarball. And it looks like it'll try to include resultant .html files causing confusion. We should figure out if we want to ship .html in the source tarballs and what we'll install if user system lacks pandoc. I think we should still install .md files regardless.

Please test and fix your work w.r.t. those requirements. I also see some unrelated changes (e.g. comments in configure.ac). Lets get those in separately.

alk commented:
Ah, also I do prefer clean history. Before we merge your work, lets rebase everything such that git history is nice.

f18m commented:
Hi Alk,
Good to see that the value of online documentation is appreciated :)

I did rebase the branch so that it now merges without conflict. I played with rebasing and I think I figured out how to stash all commits together.

Regarding "make dist": I tested and resulting tarballs build just fine with or without pandoc installed.
Regarding .md files installation, my rationale was that if you want to consult documentation locally, then you probably are going to use a browser that understands HTML, not Markdown. So I placed inclusion of .html files only under FOUND_CONVERTER_FOR_HTML conditional var.
If you prefer .md files only installed I can
 - remove if FOUND_CONVERTER_FOR_HTML / endif blocks
 - replace dist_doc_DATA += docs/something.html with  dist_doc_DATA += docs/something.md

Sounds good?

f18m commented:
Another 2 points
1) my last developments are in "docbranch" of my fork: 
   https://github.com/f18m/gperftools/tree/docbranch
now all HTML files have been converted to Markdown.

2) I'd suggest to put all png/gif files into a new folder docs/images to make it easier to browse documentation

f18m commented:
Hi Alk,
I think I've completed a review of the aesthetics of markdown documentation.
You can see the results here: 
    https://github.com/f18m/gperftools/tree/master/docs 

I think I didn't miss something, let me know if that looks OK to you.
Then I will stash all commits together so this pull request will appear as a single big commit.

Thanks,
Francesco

f18m commented:
Ping :)

alk commented:
Result looks good. Thanks. Couple more comments:

a) would be nice if configure prints warning about pandoc thingy missing

b) after I installed pandoc, make -j8 failed for me on first invocation complaining about lack of rule for readme.html or something like that

c) would be nice if we install .md docs as well. I'd consider installing html and md versions of the docs seperately. E.g. $prefix/share/doc/gperftools/html/readme.html

alk commented:
I am not able to reproduce 'make -j8' issue. Not sure why.

Installing .md and .html is good enough for now.

Couple more things. In cpu profiler format page I am seeing weird gap before  table in binary profile records section.

Also make rule for building .html files that includes sed 'inline' rewriting links is not actually 100% correct. Think of what happens when someone or something interrupts the build. Either pipe all of that in one invocation, or ensure make is aware of intermediate files (i'll smartly remove than afterwards).

I am also seeing non-working image links it tcmalloc.html. Looks like links are pointing into image/ but this is not how we install the files.

f18m commented:
Hi alk,

> Couple more things. In cpu profiler format page I am seeing weird gap before table in binary profile records section.

You're right. Fixed it.

> Also make rule for building .html files that includes sed 'inline' rewriting links is not actually 100% correct. Think of what happens when someone or something interrupts the build. Either pipe all of that in one invocation, or ensure make is aware of intermediate files (i'll smartly remove than afterwards).

Fixed by concatenating the 2 steps into 1 GNU make step

> I am also seeing non-working image links it tcmalloc.html. Looks like links are pointing into image/ but this is not how we install the files.

Good point. that required writing a custom (trivial) installation rule in "install-data-hook"...

I checked in all the changes above

alk commented:
I get this new warnings from automake:

docs/Makefile.am:52: warning: variable 'ADDITIONAL_IMAGE_SOURCES' is defined but no program or
docs/Makefile.am:52: library has 'ADDITIONAL_IMAGE' as canonical name (possible typo)

It seemingly just complains that XXX_SOURCES is commonly used to list sources for target XXX.

Another minor thing is page titles (i.e. thingy browsers reflect in tab name) are missing now.

I've noticed couple places where some italic or bold text isn't really so. I.e. section names in pprof_remote_servers.html. Or

```
CPUPROFILE_FREQUENCY=_x_
```

I think pprof_remote_servers should just be removed.

alk commented:
Another issue is that make dist doesn't put images into archive (but does put html files). Which then makes build fail.

It is nice if source tarball does have html files and we can install them when end-user system doesn't have pandoc. But not big deal if we don't.

f18m commented:
Hi alk,
I did the following changes:
 - fixed autogen warnings about ADDITIONAL_IMAGE_SOURCES
 - attempted to get HTML titles right: that requires using the Pandoc "pandoc_title_block" extension. It works and requires basically replacing the "#" char with "%". Then when using pandoc you get correct HTML. Unfortunately that is a pandoc extension and is not rendered correctly on github:

   https://github.com/f18m/gperftools/blob/master/docs/cpuprofile.md 

so honestly I don't think if it's worth using it.

 - fixed CPUPROFILE_FREQUENCY and searched for other occurrences of _sometext_ inside <pre> tag but failed to find any additional occurrence
 - removed pprof_remote_servers.md
 - I think I fixed "make dist". I tried the following steps:
   ./autogen.sh
   ./configure
   make dist
   mv gperftools-2.6.3.tar.gz ~/tmp
  cd ~/tmp
  tar xvzf gperftools-2.6.3.tar.gz
  cd gperftools-2.6.3
   ./configure
   make install

and now the final "make install"  succeeds.
There is an imperfection though: image files get installed both by the normal "make install" target into /usr/local/share/doc/gperftools/ (where they don't actually need) and by "install-data-hook" in /usr/local/share/doc/gperftools/images/ (where they must be).
Unfortunately I did not quite undertstand how to tell Automake to 
 - include files in the "dist" tarball
 - but do not install them in the "install" target
Anyway I think it's a minor issue since they are a few Kbs of images that get duplicated.

alk commented:
I am not seeing any obvious problems with current state of this change. So now is time to clean up history, rebase and get it merged at last. Thanks.

alk commented:
Sure.

sys-ptipu commented:
**:closed_book: MEV SW OD Build <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/run_od_builds/27839/display/redirect>#27839</a> result: FAILURE**
 | Manifest Name | Build Status | Jenkins Link | Artifacts | Unsuccessful Stages | Description |
 | - | - | - | - | - | - |
 | mev-ts-trunk  | FAILURE  | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-ts-trunk/59604/display/redirect>59604</a>  | <a href=https://ubit-artifactory-or.intel.com/artifactory/list/mountevans_sw_bsp-or-local/builds/official/mev-ts-trunk/od/mev-ts-trunk-od-59604> Artifacts </a> | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-ts-trunk/59604/display/redirect>IPU SDK Build</a> |  |
 | mev-trunk  | FAILURE  | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-trunk/56680/display/redirect>56680</a>  | <a href=https://ubit-artifactory-or.intel.com/artifactory/list/mountevans_sw_bsp-or-local/builds/official/mev-trunk/od/mev-trunk-od-56680> Artifacts </a> | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-trunk/56680/display/redirect>Simics_C0_ACC</a> |  |
 | mev-release | NOT AFFECTED | | | |
 | mev-ts-release | NOT AFFECTED | | | |
 | mev-ts-release-1.6.1 | NOT AFFECTED | | | |
 | mev-ts-release-bitbake | NOT AFFECTED | | | |
 | mev-ts-release-external | NOT AFFECTED | | | |
 | mev-ts-trunk-bitbake | NOT AFFECTED | | | |
 | mev-ts-trunk-bitbake-c1 | NOT AFFECTED | | | |
 | mev-ts-trunk-external | NOT AFFECTED | | | |
 | mev-upstream-idpf | NOT AFFECTED | | | |

sys-ptipu commented:
**:closed_book: MEV SW OD Build <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/run_od_builds/27894/display/redirect>#27894</a> result: FAILURE**
 | Manifest Name | Build Status | Jenkins Link | Artifacts | Unsuccessful Stages | Description |
 | - | - | - | - | - | - |
 | mev-trunk  | FAILURE  | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-trunk/56702/display/redirect>56702</a>  | <a href=https://ubit-artifactory-or.intel.com/artifactory/list/mountevans_sw_bsp-or-local/builds/official/mev-trunk/od/mev-trunk-od-56702> Artifacts </a> | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-trunk/56702/display/redirect>IPU_IMC_6_6_4</a> |  |
 | mev-ts-trunk  | SUCCESS  | <a href=https://cje-ir-prod01.devtools.intel.com/nex-ncng-ipu/job/MEV/job/builds/job/official/job/mev-ts-trunk/59633/display/redirect>59633</a>  | <a href=https://ubit-artifactory-or.intel.com/artifactory/list/mountevans_sw_bsp-or-local/builds/official/mev-ts-trunk/od/mev-ts-trunk-od-59633> Artifacts </a> |  |  |
 | mev-release | NOT AFFECTED | | | |
 | mev-ts-release | NOT AFFECTED | | | |
 | mev-ts-release-1.6.1 | NOT AFFECTED | | | |
 | mev-ts-release-bitbake | NOT AFFECTED | | | |
 | mev-ts-release-external | NOT AFFECTED | | | |
 | mev-ts-trunk-bitbake | NOT AFFECTED | | | |
 | mev-ts-trunk-bitbake-c1 | NOT AFFECTED | | | |
 | mev-ts-trunk-external | NOT AFFECTED | | | |
 | mev-upstream-idpf | NOT AFFECTED | | | |

sys-ptipu commented:
mev-ts-trunk BAT tests: [success](http://10.166.204.21/test_session?id=59188) - build 59633 - it is required if PR have BAT MEV status check as mandatory. If failed, please review Tasks in provided link to Berta. For questions please contact with ipu.ci.triage.team@intel.com
